Security self assessment for incubation application. (#42)
kubewarden/community: security-self-assesment.md

4. **Audit Scanner**: cronjob that independently inspects cluster resources to
   identify policy violations, operating separately from the admission control
   flow. It achieves this by creating synthetic requests that are evaluated by
   the PolicyServers. The results get saved into PolicyReports CRs.

5. **Kubernetes API Server**: interacts with the PolicyServer through
   ValidatingWebhookConfiguration and MutatingWebhookConfiguration
   objects.

6. **OCI Registry**: stores and distributes Kubewarden policies as OCI artifacts,
   separate from the cluster environment.
7. **kwctl** CLI utility: command-line utility to interact with policies
   outside of the cluster. It uses the same libraries as the policy-server
   binary. It:

   - Pulls, pushes policies from and to OCI registries.
   - Verifies policy integrity and authenticity using Sigstore when downloading
     those policies.
   - Inspects the policy metadata (policies binaries are annotated with
     metadata). Annotates policy binaries with metadata. Metadata includes the
     types of the Kubernetes resources the policy targets, and the policy
     capabilities (mutating, context-aware, etc).

>    - Runs policies outside of the cluster. Useful for developing, benchmarking

Author: ```suggestion
   - Runs policies outside the cluster. Useful for developing, benchmarking
```

   with that PolicyServer.

7. Policy execution for usual Admission Control:

   1. Kubernetes API Server sends admission requests to the policy-server
      binary via webhooks.
   2. policy-server validates the request format and authenticity.
   3. policy-server executes relevant WebAssembly policies in a sandboxed
      environment. if the policy is a mutating one, the end result is a changed
      admission request.
   4. policy-server returns admission decision to the API Server, with the
      decision to accept or reject the request.

8. Policy execution for context-aware calls:

   It is the same process as a normal execution, with the difference that the
   policy-server performs queries to the API server to learn about the state of
   the cluster. This queries are done with a Kubernetes client, and via the
   ServiceAccount that the PolicyServer was configured with. By default,
   readonly access is provided.
   In addition, each policy has a `spec.contextAwareResources` field that lists
   which resources from the cluster the policy is allowed to query about. In
   case that a policy Wasm module tries to access resources that the Policy CR

>    didn't declare in its `spec.contextAwareResources`, said query is blocked

Author: Your original is absolutely correct, maybe just that 'that' is simpler/clearer. 
```suggestion
   didn't declare in its `spec.contextAwareResources`, that query is blocked
```

      resources and the policies that apply to those resources.
   3. It queries the cluster via its default ServiceAccount, which by
      default is bound to the ClusterRoles `view` (read-only of most objects
      besides Secres, Roles, or RoleBindings) and `audit-scanner-cluster-role`
      (read-write access to Kubewarden resources and PolicyReports). The SA is
      configurable.
   4. The audit-scanner binary creates [synthetic
      requests](https://docs.kubewarden.io/explanations/audit-scanner), and
      sends them to the policy servers.
      Some [limitations
      apply](https://docs.kubewarden.io/explanations/audit-scanner/limitations)
      (e.g: policies relaying on external data, or user and user group info).
   5. The audit-scanner collects all the results and saves them in
      PolicyReports from the openreports.io project, deleting old PolicyReports
      on each run first.
      This ensures that PolicyReports results are as up-to-date as possible.

10. Secure Communication:

    - Kubewarden Controller generates and manages TLS certificates.
    - Components use these certificates for encrypted communication.

>     - If desired, one can enable mutual TLS or use NetworkPolicies and a CNI to

Author: ```suggestion
    - If needed, one can enable mutual TLS or use NetworkPolicies and a CNI to
```

    - UI presents compliance status and policy violations in a secure,
      read-only interface.

#### Goals

Kubewarden intends to:

- Be the Universal Policy Engine for Kubernetes by simplifying adoption of
  Policy As Code in cloud-native environments. Provide security hardening,
  continous compliance auditing, and resource optimization for Kubernetes
  clusters.
- Enable writing policies in any programming language that generates
  WebAssembly binaries.
  - Allow reuse of policies from other policy engines without rewriting them.
    For example reusing Rego policies from OPA & Gatekeeper, or using our
    experimental `kyverno-dsl-policy` policy.
  - Distribute policies using OCI compliant registries.
  - Provide a stable API for policy development, and SDKs for popular languages.
- Enhance security by verifying policies using SLSA, as well as Kubernetes
  resources.
- Cater to various personas in an organization, including Policy Users, Policy
  Developers, Policy Distributors, Cluster Operators, and Kubewarden Integrators.
- Foster transparency, collaboration, and improvement through CNCF membership
  and an open-source community.

> - Maintain a corpus of useful policies, shipped via ArtifactHub.

Author: ```suggestion
- Maintain a collection of useful policies, shipped via ArtifactHub.
```

#### OpenSSF Best Practices

Kubewarden passes the [Open SSF Best Practices](https://www.bestpractices.dev/en):
https://www.bestpractices.dev/en/users/23101

Kubewarden is also in
[CLOMonitor](https://clomonitor.io/projects/cncf/kubewarden), and we strive to
have 100% coverage.

#### Adopters and Case Studies

We maintain a list of adopters in the [ADOPTERS.md](./ADOPTERS.md) file of this
repository.

Example case studies:

- Case A: as a Kubernetes operator, I want to ensure my cluster is safe and
  compliant.

  I deploy Kubewarden and its default configuration with its
  `kubewarden-defaults` Helm chart, in the `kubewarden` Namespace. This deploys a
  default PolicyServer and recommended ClusterAdmissionPolicies in the
  `kubewarden` namespace, solely under the control of the Kubernetes operator.

  As an operator, I can add more PolicyServers under a managed Namespace
  (such as `kubewarden`) to distribute load and fault tolerance.

  The operator can deploy more ClusterAdmissionPolicies and ClusterAdmissionPolicyGroups
  that check the totality of the Kubernetes resources, for any type of

>   operation (CREATE, UPDATE, DELETE, etc). This ensures operations into the

Author: I'm not sure about my suggestion here. I don't necessarily know what I'm talking about here. The etc. to me implies many more operations. So, maybe be just spell out the 6. I think it's 6. I'm using this as reference: https://kubernetes.io/docs/reference/using-api/api-concepts/#single-resource-api
```suggestion
  operation (GET, CREATE, UPDATE, PATCH, DELETE, PROXY). This ensures operations into the
```

  cluster are safe and compliant. This includes security, compliance (to
  industry standards or company regulations), resource optimization (via
  mutating policies), governance of Kubernetes environments (via labels and
  naming conventions), best practices, image verification, etc.

>   The security expectations change in the future, and that was ok to have
>   deployed in the cluster isn't anymore. Yet Kubewarden already accepted those
>   operations in the cluster. What the operator can do is to deploy the Audit
>   Scanner feature, a CronJob that runs periodically and evaluates the existing
>   resources in the cluster. This ensures the cluster is safe and compliant even
>   with the pass of time.

Author: ```suggestion
  There will be changes in security expectations in the future. 
  What was previously correct to be deployed in the cluster may no longer be correct. 
  Yet Kubewarden has already accepted those operations in the cluster. In these situations the operator can deploy the Audit
  Scanner feature, a CronJob that runs periodically and evaluates the existing
  resources in the cluster. This ensures the cluster is safe and compliant even
  over time.
```

  The operator can configure some or all the policies in `monitor` mode instead
  of `protect` mode, to learn from the state of the cluster without blocking operations.

  As usual, the operator can receive information from policies and the Kubewarden stack
  by consuming logs and OpenTelemetry information for metrics and tracing.

- Case B: As a Kubernetes operator, I want to provide a framework to my
  Kubernetes users so they can self-service in their Namespaces.

  As an operator, I deploy Kubewarden as in Case A for a set of policies of my choosing.
  This provides me with safe baseline in the cluster, that other users cannot evade.

  In addition to Case A, I have different personas per Namespaces: perhaps
  teams, team administrators, test deployments, etc.
  I allow each Namespace administrator to self-service by letting them deploy
  PolicyServers in their Namespace, along with namespaced AdmissionPolicies and

>   AdmissionPolicyGroups. These arquitecture means that they are in control of

Author: ```suggestion
  AdmissionPolicyGroups. This architecture means that they are in control of
```

  first-class support.

  Kubewarden policies can be as simple as desired, or complex
  [context-aware](https://docs.kubewarden.io/explanations/context-aware-policies)
  policies. Context-aware policies can also be used to interface with separate
  workloads (for example, to obtain information from an image scanner
  long-running service).

- Case D: As a system integrator, I want to re-use Kubewarden as part of my
  security and compliante solution, and re-use other solutions by including
  them in Kubewarden.

  As a system integrator, I can reuse parts of Kubewarden, such as the
  `policy-server`, to police resources inside of the Kubernetes cluster, or
  outside of the cluster via the ["raw
  policies"](https://docs.kubewarden.io/howtos/raw-policies) feature.

  The system integrator can chose to deploy the `kubewarden-controller` or
  manage the CRDs on their own, and can chose to deploy or scale the Audit
  Scanner as needed.

  I can create new components, for example an image scanner, and interface with
  it via a context-aware policy, without having a monolithic implementation in
  a Kubernetes controller.

#### Related Projects / Vendors

> We maintain a comparison documentation page with comparisons. Currently,

Author: ```suggestion
We maintain a comparison documentation page. Currently,
```

> it contains a [comparison with OPA Gatekeeper](https://docs.kubewarden.io/explanations/comparisons/opa-comparison).

Author: ```suggestion
it compares [Kubewarden and OPA Gatekeeper](https://docs.kubewarden.io/explanations/comparisons/opa-comparison).
```
